sql-plugin/src/main/scala/com/nvidia/spark/rapids/ParquetCachedBatchSerializer.scala

```diff
@@ -615,6 +615,184 @@ protected class ParquetCachedBatchSerializer extends GpuCachedBatchSerializer {
     }
   }
 
+  /**
+   * Read the data from an input file and convert the result to rows. The data will not be
+   * transformed in any way. It will just be read.
+   */
+  private class PcbFileToRowsIterator(inputFile: ByteArrayInputFile,
+      cacheAttributes: Seq[Attribute],
+      hadoopConf: Configuration,
+      options: ParquetReadOptions) extends Iterator[InternalRow] with AutoCloseable {
+    import org.apache.parquet.io.ColumnIOFactory
+
+    private[this] val parquetFileReader = ParquetFileReader.open(inputFile, options)
+    private[this] val parquetSchema = parquetFileReader.getFooter.getFileMetaData.getSchema
+    private[this] val columnIO = new ColumnIOFactory().getColumnIO(parquetSchema)
+
+    private[this] var pages: org.apache.parquet.column.page.PageReadStore = _
+    private[this] var rowsLeftInRowGroup: Long = _
+    private[this] var recordReader: org.apache.parquet.io.RecordReader[InternalRow] = _
+    // Prime the data with the first row group
+    goToNextRowGroup
+
+    private[this] def goToNextRowGroup: Unit = {
+      pages = parquetFileReader.readNextRowGroup()
+      if (pages == null) {
+        rowsLeftInRowGroup = 0
+        recordReader = null
+      } else {
+        rowsLeftInRowGroup = pages.getRowCount
+        recordReader = columnIO.getRecordReader(pages, new ParquetRecordMaterializer(parquetSchema,
+          cacheAttributes.toStructType,
+          new ParquetToSparkSchemaConverter(hadoopConf), None /*convertTz*/ ,
+          LegacyBehaviorPolicy.CORRECTED))
+      }
+    }
+
+    override def hasNext: Boolean = pages != null
+
+    override def next(): InternalRow = {
+      if (!hasNext) {
+        throw new NoSuchElementException()
```

Comment:
nit: While we are here it might make sense to add a message to this exception.

Reply:
What message would you have me add? I have no more information for the user on why they walked past the end of the iterator.  They will know from the stack trace that this came from the ParquetCachedBatchSerializer. I don't know what it would add.

If you look through our code the vast majority of the NoSuchElementExceptions that we throw do not have a message, and the messages that are included don't really help most of the time.